From 8977ec97a49c0bb0d063df22d30153eb02370f40 Mon Sep 17 00:00:00 2001
From: BellSoft <info@bell-sw.com>
Date: Sat, 10 Oct 2026 19:10:46 +0000
Subject: [PATCH] feat(purl): Add more PURL-to-ecosystem mappings

This PR adds more mappings for the following PURL types to their
corresponding OSV ecosystems:
- alpaquita
- bellsoft-hardened-containers

(google/osv-scanner#2862 by Ildar Mulyukov)
---
 internal/utility/purl/purl_to_package.go      |  8 +++++--
 internal/utility/purl/purl_to_package_test.go | 22 +++++++++++++++++++
 2 files changed, 28 insertions(+), 2 deletions(-)

diff --git a/internal/utility/purl/purl_to_package.go b/internal/utility/purl/purl_to_package.go
index 9e706a8..4994f0d 100644
--- a/internal/utility/purl/purl_to_package.go
+++ b/internal/utility/purl/purl_to_package.go
@@ -12,7 +12,11 @@ import (
 // used like so: purlEcosystems[PkgURL.Type][PkgURL.Namespace]
 // * means it should match any namespace string
 var purlEcosystems = map[string]map[string]osvconstants.Ecosystem{
-	"apk":      {"alpine": osvconstants.EcosystemAlpine},
+	"apk": {
+		"alpaquita":                    osvconstants.EcosystemAlpaquita,
+		"alpine":                       osvconstants.EcosystemAlpine,
+		"bellsoft-hardened-containers": osvconstants.EcosystemBellSoftHardenedContainers,
+	},
 	"cargo":    {"*": osvconstants.EcosystemCratesIO},
 	"composer": {"*": osvconstants.EcosystemPackagist},
 	"conan":    {"*": osvconstants.EcosystemConanCenter},
@@ -100,7 +104,7 @@ func ToPackage(purl string) (models.PackageInfo, error) {
 		case osvconstants.EcosystemMaven:
 			// Maven uses : to separate namespace and package
 			name = parsedPURL.Namespace + ":" + parsedPURL.Name
-		case osvconstants.EcosystemDebian, osvconstants.EcosystemAlpine, osvconstants.EcosystemUbuntu:
+		case osvconstants.EcosystemAlpaquita, osvconstants.EcosystemAlpine, osvconstants.EcosystemBellSoftHardenedContainers, osvconstants.EcosystemDebian, osvconstants.EcosystemUbuntu:
 			// Debian and Alpine repeats their namespace in PURL, so don't add it to the name
 			name = parsedPURL.Name
 		default:
diff --git a/internal/utility/purl/purl_to_package_test.go b/internal/utility/purl/purl_to_package_test.go
index 8f7dd6d..13e7be4 100644
--- a/internal/utility/purl/purl_to_package_test.go
+++ b/internal/utility/purl/purl_to_package_test.go
@@ -130,6 +130,28 @@ func TestPURLToPackage(t *testing.T) {
 				Ecosystem: string(osvconstants.EcosystemAlpine),
 			},
 		},
+		{
+			name: "valid_PURL_alpaquta",
+			args: args{
+				purl: "pkg:apk/alpaquita/busybox@1.37.0-r37?arch=x86_64&distro=alpaquita-stream",
+			},
+			want: models.PackageInfo{
+				Name:      "busybox",
+				Version:   "1.37.0-r37",
+				Ecosystem: string(osvconstants.EcosystemAlpaquita),
+			},
+		},
+		{
+			name: "valid_PURL_bellsoft-hardened-containers",
+			args: args{
+				purl: "pkg:apk/bellsoft-hardened-containers/apk-tools@3.0.6-r0?arch=x86_64&distro=bellsoft-hardened-containers-stream",
+			},
+			want: models.PackageInfo{
+				Name:      "apk-tools",
+				Version:   "3.0.6-r0",
+				Ecosystem: string(osvconstants.EcosystemBellSoftHardenedContainers),
+			},
+		},
 		{
 			name: "invalid_PURL",
 			args: args{
-- 
2.43.0

